clients/rust/src/input_handling.rs

```diff
@@ -102,6 +103,67 @@ async fn resolved_input_message_content_to_client_input_message_content(
     }
 }
 
+/// Since we store the input in the database in the form of ResolvedInput but without e.g. images inside,
+/// we need to reresolve the input when we retrieve it from the database.
+/// Resolves images in place.
+pub async fn reresolve_input_for_fine_tuning(
+    input: &mut ResolvedInput,
+    client: &Client,
+) -> Result<(), TensorZeroError> {
+    let mut image_fetch_tasks = Vec::new();
+
+    for (message_index, message) in input.messages.iter_mut().enumerate() {
+        // First pass: identify images to fetch and collect tasks
+        for (content_index, content) in message.content.iter_mut().enumerate() {
+            if let ResolvedInputMessageContent::Image(image_with_path) = content {
+                if image_with_path.image.data.is_none() {
+                    let storage_path = image_with_path.storage_path.clone();
+                    let fut = async move {
+                        let result = fetch_image_data(storage_path, client).await;
```

Comment:
This can be:
```rust
let result = fetch_image_data(storage_path, client).await?;
Ok((message_index, content_index, result))
```

This will let you use `try_join_all` to bail out early on an error